genmc-sys/build.rs

```diff
@@ -0,0 +1,200 @@
+                    println!("cargo::error=Cannot open GenMC repo at path '{GENMC_LOCAL_PATH_STR}': {open_err:?}");
+                    println!("cargo::error=Cannot clone GenMC repo from '{GENMC_GITHUB_URL}': {clone_err:?}");
+                    std::process::exit(1);
+                }
+            }
+        });
+
+        let statuses = repo.statuses(None).expect("should be able to get repository status");
+        if !statuses.is_empty() {
+            println!(
+                "cargo::error=Downloaded GenMC repository at path '{GENMC_DOWNLOAD_PATH_STR}' has been modified"
+            );
+            for entry in statuses.iter() {
+                println!(
+                    "cargo::error=  {} is {:?}",
+                    entry.path().unwrap_or("unknown"),
+                    entry.status()
+                );
+            }
+            println!(
+                "cargo::error=This repository should only be modified by the 'genmc-sys' build script to load a specific commit ('{GENMC_COMMIT}')"
+            );
+            println!(
+                "cargo::error=HINT: For local development, place a GenMC repository in the path '{GENMC_LOCAL_PATH_STR}'"
+            );
+            std::process::exit(1);
+        }
+
+        // Check if there are any updates:
+        let commit = if let Ok(oid) = Oid::from_str(GENMC_COMMIT)
+            && let Ok(commit) = repo.find_commit(oid)
+        {
+            commit
+        } else {
+            match repo.find_remote("origin") {
+                Ok(mut remote) =>
+                    match remote.fetch(&[GENMC_COMMIT], None, None) {
+                        Ok(_) =>
+                            println!(
+                                "cargo::warning=Successfully fetched commit '{GENMC_COMMIT:?}'"
+                            ),
+                        Err(e) => panic!("Failed to fetch from remote: {e}"),
+                    },
+                Err(e) => println!("cargo::warning=Could not find remote 'origin': {e}"),
+            }
+            let oid = Oid::from_str(GENMC_COMMIT).unwrap();
+            repo.find_commit(oid).unwrap()
+        };
+
+        // Set the repo to the correct branch:
+        checkout_commit(&repo, GENMC_COMMIT);
+
+        let head_commit = repo.head().unwrap().peel_to_commit().unwrap();
+        assert_eq!(head_commit.id(), commit.id());
+        println!("cargo::warning=Successfully set checked out commit {head_commit:?}");
+
+        genmc_download_path
+    }
+
+    fn checkout_commit(repo: &Repository, refname: &str) {
+        let (object, reference) = repo.revparse_ext(refname).expect("Object not found");
+
+        repo.checkout_tree(&object, None).expect("Failed to checkout");
+
+        match reference {
+            // `gref` is an actual reference like branches or tags.
+            Some(gref) => repo.set_head(gref.name().unwrap()),
+            // This is a commit, not a reference.
+            None => repo.set_head_detached(object.id()),
```

Comment:
On the terminal I could just do `git checkout $commit` or `git reset --hard $commit`. Why does this need a match? Is there no command that always works?